Add guard to dt negative case and break the controller for safety

alcantara09 opened the pull request:
Hi,

This PR adds a guard to the case when the time step on the diff_drive_controller is computed as negative due to some error coming from the hardware interface, as in https://github.com/ros-controls/ros_controllers/issues/561

At my company we are experiencing the same backward movement behavior but less consistent and less predicable. This safe guards the controller to this type of error, what can be common given we are also experiencing the same issue.

Commit:
guard dt negative case and break the controller

matthew-reynolds reviewed (changes requested):
We should guard on `period` being negative, which corresponds to `cmd_dt` rather than `dt`, Could you also duplicate this over to `ackermann_steering_controller` and `four_wheel_steering_controller`? Those controllers all use the period in their `SpeedLimiters`, which is where this particularly nasty behaviour is coming from.

The other controllers only use the period for updating PIDs, which will still have some odd effects but shouldn't be nearly as big of a deal.

I will add, though, that if you're seeing this issue then the real solution is to fix your `ControllerManager::Update()` to use a monotonic time source. Adding warnings here is a bit of a fall back and diagnostic, it shouldn't be relied on to provide good behaviour. Passing in a negative period isn't something the framework is designed to handle, since it makes no sense.

alcantara09 commented:
> We should guard on `period` being negative, which corresponds to `cmd_dt` rather than `dt`, Could you also duplicate this over to `ackermann_steering_controller` and `four_wheel_steering_controller`? Those controllers all use the period in their `SpeedLimiters`, which is where this particularly nasty behaviour is coming from.
> 
> The other controllers only use the period for updating PIDs, which will still have some odd effects but shouldn't be nearly as big of a deal.
> 

Ok, I will do the following changes, no problem.

> I will add, though, that if you're seeing this issue then the real solution is to fix your `ControllerManager::Update()` to use a monotonic time source. Adding warnings here is a bit of a fall back and diagnostic, it shouldn't be relied on to provide good behaviour. Passing in a negative period isn't something the framework is designed to handle, since it makes no sense.

You are absolutely correct, we are debugging our update to correct the problem. But given we were not the only ones to hit this issue, I thought it would be useful to have the guard

alcantara09 commented:
Hi,

I did the asked changes. 

However I have one question still. I was getting a lot of negative values for `dt`. What is the meaning of this? should we add a `std::abs` to check the timeout with a positive value?

alcantara09 commented:
Hi,

I was just testing the entire chain and I just realized that setting curr_cmd to zero doesn't guard against the bug because it will be still inserted on the limiter. 

I see two options then:
1. We can set curr_cmd to zero and skip the limiter step
2. We can set curr_cmd to zero and set cmd_dt to a positive value (The question then would be which value)

An alternative would be to throw an exception

What is your opinion on that? 

(The unit tests are running normally locally, I do not understand why it failed on the pipeline)

matthew-reynolds commented:
I think we probably want to just set the `cmd_dt` to 0. That should force the acceleration and jerk to be 0 - In other words, maintain the same vel as the prev update. Not perfect, but I think that's the safest solution.

Skipping the limiter step opens up a hole for a dangerous command to slip through, which I don't think is acceptable.

Commit:
set cmd_dt to zero in case period is negative

alcantara09 commented:
Fair enough, I did the changes

matthew-reynolds commented on ackermann_steering_controller/src/ackermann_steering_controller.cpp:
```diff
@@ -321,7 +321,13 @@ namespace ackermann_steering_controller{
 
     // Limit velocities and accelerations:
-    const double cmd_dt(period.toSec());
+    const double cmd_dt = (period.toSec() > 0) ? period.toSec() : 0.0;
+    if(cmd_dt < 0)
+    {
+      ROS_ERROR("Invalid time interval, delta time cannot be negative");
+      curr_cmd.lin = 0.0;
+      curr_cmd.ang = 0.0;
+    }
```
Your proposed change won't issue the warning, and we shouldn't touch the curr_cmd.

Something like this makes more sense.

```suggestion
    double cmd_dt(period.toSec());
    if (cmd_dt < 0.0)
    {
      ROS_ERROR("Invalid time interval, delta time cannot be negative");
      cmd_dt = 0.0;
    }
```

matthew-reynolds reviewed (changes requested):
Since we don't have tests to verify this behaviour, please test the changes locally to ensure the PR has the desired effect. Your last commit doesn't appear to satisfy the intended behaviour.

Please also retarget this PR to the `noetic-devel` branch. It can then be backported from there to Melodic.

Commit:
remove setting curr_cmd to zero when cmd_dt is negative to avoid abrupt changes in control

bmagyar commented:
Any update on this @alcantara09 ? I'd like to merge a working solution if you have one. The current errors printed are not bad to start but if you could set the values to 0.0 on top of reporting the error that'd make things a little safer.

Commit:
correct negative time check for error log

Since we are already setting cmd_dt to zero in the case of negative
period, we need to check the period itself in order to log the error

alcantara09 commented:
Hi @bmagyar and @matthew-reynolds,

I am forcing cmd_dt to zero when we have a negative period. 
This forces all jerk and acceleration to be zero and then we are forwarding last cmd as current cmd, as it was agreed as expected behavior. Only on the jerk calculation, `v = dv0 + da`, that the contributtion of the previous velocities still can alter the current cmd. What I do not believe should be set to zero.

I believe the code is in conformity with the issues raised by you, otherwise could you elaborate a little more the aspect that I am still missing so I can change?
